server/datastore/mysql/software.go

```diff
@@ -1581,61 +1594,91 @@ func (ds *Datastore) SyncHostsSoftware(ctx context.Context, updatedAt time.Time)
 func (ds *Datastore) ReconcileSoftwareTitles(ctx context.Context) error {
 	// TODO: consider if we should batch writes to software or software_titles table
 
-	// ensure all software titles are in the software_titles table
-	upsertTitlesStmt := `
-INSERT INTO software_titles (name, source, browser)
-SELECT DISTINCT
-	name,
-	source,
-	browser
-FROM
-	software s
-WHERE
-	NOT EXISTS (SELECT 1 FROM software_titles st WHERE (s.name, s.source, s.browser) = (st.name, st.source, st.browser))
-ON DUPLICATE KEY UPDATE software_titles.id = software_titles.id`
-	// TODO: consider the impact of on duplicate key update vs. risk of insert ignore
-	// or performing a select first to see if the title exists and only inserting
-	// new titles
-
-	res, err := ds.writer(ctx).ExecContext(ctx, upsertTitlesStmt)
-	if err != nil {
-		return ctxerr.Wrap(ctx, err, "upsert software titles")
-	}
-	n, _ := res.RowsAffected()
-	level.Debug(ds.logger).Log("msg", "upsert software titles", "rows_affected", n)
-
-	// update title ids for software table entries
-	updateSoftwareStmt := `
-UPDATE
-	software s,
-	software_titles st
-SET
-	s.title_id = st.id
-WHERE
-	(s.name, s.source, s.browser) = (st.name, st.source, st.browser)
-	AND (s.title_id IS NULL OR s.title_id != st.id)`
+	return ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+		// ensure all software titles are in the software_titles table
+		upsertTitlesStmt := `
+INSERT INTO software_titles (name, source, browser, bundle_identifier)
+SELECT
+    name,
+    source,
+    browser,
+    bundle_identifier
+FROM (
+    SELECT DISTINCT
+        name,
+        source,
+        browser,
+        bundle_identifier
+    FROM
+        software s
+    WHERE
+        NOT EXISTS (
+            SELECT 1 FROM software_titles st
+            WHERE s.bundle_identifier = st.bundle_identifier
+        )
+        AND bundle_identifier != ''
```

Comment:
`software.bundle_identifier`  can be `NULL` looks like?

```suggestion
        AND COALESCE(bundle_identifier, '') != ''
```

Reply:
this is super confusing, I could use your help considering if the approach in general is good.

- `software` has `NOT NULL DEFAULT ''` for `bundle_identifier`
- `software_titles` has `DEFAULT NULL` for `bundle_identifier` (for the reasons we chatted above)

this is the `bundle_identifier` for `software` so it should really be, for clarity:

```suggestion
        AND s.bundle_identifier != ''
```

but the main point is that the two tables have a column with identical name, for the same thing, but different schema.

Reply:
At least on my local env (I may be missing something?) `software.bundle_identifier` can be NULL? It does default to `''`, but if specified explicitly it could have NULL values (might be something we'd never encounter, but since we tend to see all kinds of things in the wild...).

```
Create Table: CREATE TABLE `software` (
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `name` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL,
  `version` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '',
  `source` varchar(64) COLLATE utf8mb4_unicode_ci NOT NULL,
  `bundle_identifier` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT '',
  `release` varchar(64) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '',
  `vendor_old` varchar(32) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '',
  `arch` varchar(16) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '',
  `vendor` varchar(114) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '',
  `browser` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '',
  `extension_id` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '',
  `title_id` int(10) unsigned DEFAULT NULL,
  `checksum` binary(16) NOT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `idx_software_checksum` (`checksum`),
  KEY `software_listing_idx` (`name`,`id`),
  KEY `software_source_vendor_idx` (`source`,`vendor_old`),
  KEY `title_id` (`title_id`),
  KEY `idx_sw_name_source_browser` (`name`,`source`,`browser`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
```

Reply:
good point, thanks! I completely missed that possibility, adjusting now